### PR TITLE
Fair-2913: Worked example for create/update cycle of datasets

### PR DESCRIPTION
**Pull request by seantc**
This adds a command to demonstrate update cycle of datasets `fair-api-datasets-update.py`. Adds documentation to describe how create / update cycle should work.

It also adds a --dry-run option to `fair-api-datasets-create.py` and `fair-api-datasets-update.py` commands to allow for experimentation.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -287,6 +287,20 @@ Examples:
...
+
+This section describes how the update cycle in FAIR works.
+
+To demonstrate how to update a dataset, the `fair-api-datasets-update.py` helper command can be used:
+
+1. Start by creating a JSON file conforming to the metadata specification set out in [Create a Dataset Entry](#). We will call this `dataset.json`. Post this to the API and make a note of the dataset_code.
+2. Make a change to the metadata in `dataset.json`, for example, add or remove a dictionary from the `dictionaries` list, or change some values in the `catalogue` attributes.
+3. Now call the `fair-api-datasets-update.py` command with `--dry-run` enabled. This will calculate differences and print the appropriate PATCH call that should be used to update the dataset. Running witout --dry-run will apply the change to the database:
+```
+python3 fair-api-datasets-update.py ~/project/scratch/helper_post_patch/dataset.json --dry-run
```
Comment (RossBarnie) [inline README.md:300]: It's stated further up in the README that python is assumed to be python 3, so I think we should be consistent with other commands and use `python` over `python3`.

File: datasets/command_base.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+if 'FAIR_API_TOKEN' not in os.environ:
+    print('Please add FAIR_API_TOKEN to the environment')
+    exit(1)
+
+if 'FAIR_API_ENDPOINT' not in os.environ:
+    print('Please add FAIR_API_ENDPOINT to the environment')
+    exit(1)
+
+DRY_RUN = True if (len(sys.argv) > 2 and sys.argv[2] == '--dry-run') else False
```
Comment (RossBarnie) [inline datasets/command_base.py:15]: Use the condition over `if true; return true; else; return false`, ie:

```python
DRY_RUN = len(sys.argv) > 2 and sys.argv[2] == '--dry-run'
```

File: datasets/diff_helper.py [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+    return diff
+
+  def sanitize_dict(original):
+    modified = original.copy()
+    for key in original:
+      if key not in DICTIONARY_FIELDS or original[key] == None:
+        modified.pop(key)
+    return modified
+
+  def is_equal(original, data):
```
Comment (RossBarnie) [inline datasets/diff_helper.py:71]: I don't think we need this method as it's a wrapper around the standard equality check.

File: datasets/diff_helper.py [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+  def changed_dictionaries(current, previous, codes):
+    dictionaries = []
+    for code in codes:
+      currentDictionary = DiffHelper.find_by_code(current, code)
+      previousDictionary = DiffHelper.find_by_code(previous, code)
+      previousDictionary = DiffHelper.sanitize_dict(previousDictionary)
+      if DiffHelper.is_equal(currentDictionary, previousDictionary):
+        continue
+
+      if not DiffHelper.is_equal(currentDictionary.get('fields'), previousDictionary.get('fields')) or not DiffHelper.is_equal(currentDictionary.get('lookups'), previousDictionary.get('lookups')):
```
Comment (RossBarnie) [inline datasets/diff_helper.py:108]: This line proves my point about the is_equal method.

With DiffHelper.is_equal:
```python
if not DiffHelper.is_equal(currentDictionary.get('fields'), previousDictionary.get('fields'))
```

With standard boolean operation:
```python
if currentDictionary.get('fields') != previousDictionary.get('fields')
```

**Review by RossBarnie — COMMENTED** [review]
A few minor tweaks but changes overall give some much needed organisation.

File: datasets/diff_helper.py (hunk 3 above)
Comment (seantc) [reply datasets/diff_helper.py:71]: Yeah not quite sure why this was wrapped!

**Review by RossBarnie — APPROVED** [review]

**Pull request merged**